### PR TITLE
test(gmpe): initial coverage and migrating tests to hardhat

### PR DESCRIPTION
**Pull request by deanamiel**
Initial GMPE tests, using hardhat in conjunction with waffle to detect custom errors

### CODE REVIEW

File: hardhat.config.js [hunk 1]
```diff
@@ -1,5 +1,4 @@
-require('hardhat-gas-reporter');
-require('solidity-coverage');
+require('@nomicfoundation/hardhat-toolbox');
```
Comment (milapsheth) [inline hardhat.config.js:1]: :+1:

File: test/GMP.js [hunk 2]
```diff
@@ -5,41 +5,34 @@ const {
   Contract,
   utils: { defaultAbiCoder, keccak256, id },
 } = require('ethers');
-const { deployContract, MockProvider, solidity } = require('ethereum-waffle');
-chai.use(solidity);
 const { expect } = chai;
+const { ethers } = require('hardhat');
 
 const ADDRESS_ZERO = '0x0000000000000000000000000000000000000000';
```
[the remark this answers is not in the document]
Comment (deanamiel) [reply test/GMP.js:11]: you're right, the import is implicit, I'll remove from each test file

File: test/ConstAddressDeployer.js [hunk 3]
```diff
@@ -1,27 +1,40 @@
...
+  let deployerFactory;
   let deployer;
   const name = 'test';
   const symbol = 'test';
   const decimals = 16;
 
+  before(async () => {
+    wallets = await ethers.getSigners();
+    deployerWallet = wallets[0];
+    userWallet = wallets[1];
```
[the remark this answers is not in the document]
Comment (deanamiel) [reply test/ConstAddressDeployer.js:28]: agreed, I'll make this correction in each test file

File: test/ConstAddressDeployer.js [hunk 4]
```diff
@@ -1,27 +1,40 @@
...
 const BurnableMintableCappedERC20 = require('../artifacts/contracts/test/ERC20MintableBurnable.sol/ERC20MintableBurnable.json');
 const BurnableMintableCappedERC20Init = require('../artifacts/contracts/test/ERC20MintableBurnableInit.sol/ERC20MintableBurnableInit.json');
 
 describe('ConstAddressDeployer', () => {
-  const [deployerWallet, userWallet] = new MockProvider().getWallets();
+  let wallets;
+  let deployerWallet;
+  let userWallet;
+
+  let deployerFactory;
```
[the remark this answers is not in the document]
Comment (deanamiel) [reply test/ConstAddressDeployer.js:19]: I use all the factory variables once to instantiate them in the before() hook and then again to deploy contracts in the beforeEach() hook so I think they need to be global to be accessible in both.

File: test/GMP.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (deanamiel) [reply test/GMP.js:11]: when I remove it all tests still run smoothly but now there's a lint error each time "ethers" is referenced in the code

**Review by milapsheth — APPROVED** [review]

**Review by re1ro — CHANGES_REQUESTED** [review]
Could you please rename `DummyProxy` into `InvalidExpressProxy`  and the same for registry?

**Review by re1ro — APPROVED** [review]

**Pull request merged**